Tests/Realm.Tests/Database/PrimaryKeyTests.cs

```diff
@@ -192,6 +195,70 @@ private RealmObjectBase FindByPKGeneric(Type type, object primaryKeyValue, PKTyp
             return (RealmObjectBase)genericMethod.MakeGenericMethod(type).Invoke(_realm, new[] { primaryKeyValue });
         }
 
+        [Test]
+        public void RealmFind_WhenPKIsIntAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<IntPrimaryKeyWithValueObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsCharAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyCharObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsByteAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyByteObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsInt16AndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyInt16Object>();
+
+        [Test]
+        public void RealmFind_WhenPKIsInt32AndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyInt32Object>();
+
+        [Test]
+        public void RealmFind_WhenPKIsInt64AndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyInt64Object>();
+
+        [Test]
+        public void RealmFind_WhenPKIsStringAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyStringObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsGuidAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyGuidObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsObjectIdAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyObjectIdObject>();
+
+        // In Find<T> anything that is a primitive type (according to Realm's view) is converted to long?.
+        // FindCore<T> converts its input in RealmValue. Which in turn treats long as PritiveValue of type Int.
+        // Without this pass it'd be impossible to match any PK that's of primitive type, since a PK defined as e.g. short would always be
+        // compared against an int
+        private static Type ConvertToRealmValue(Type toConvert)
```

Comment:
I don't think of this as a good way to achieve what's needed, but after various attempts I folded back to a simple solution.
I'm confident you can advice for better. At least this allowed to write a working version of the tests following your feedback.

Reply:
This should better be renamed to `GetDatabaseType` as initially I was confused how is that related to RealmValue (which is our implementation of Mixed).

It can be simplified a little with a few suggestions:

1. `nativeTypes` is the incorrect variable name here because those types are not really native - instead it should be something like `intBackedTypes` or `intConvertibleTypes`.
2. You don't need it to be a `ReadOnlyCollection` - a simple array works just fine.
3. You can use LINQ's `Contains` to avoid doing the manual loop.
4. For int-backed types, you should return `typeof(long)` as that is more semantically correct. This will require to change `123` in `_primaryKeyValues` to `123L`.
5. If the type is not stored as an Int64 internally, you can return the argument rather than returning `null`.